### PR TITLE
fix for the while_let_loop!

### LINKED ISSUE #359: `while_let_loop` fails to lint if last expression in `loop` is a block
Another `while_let_loop` issue courtesy of the [n-body benchmark game](http://benchmarksgame.alioth.debian.org/u64q/program.php?test=nbody&lang=rust).

This compiles with no warnings:
```
fn no_warn<T>(slice: &[T]) {
    let mut iter = slice.iter();
    loop {
        let _ = match iter.next() {
            Some(ele) => ele,
            None => break
        };
        {}
    }
}
```

This properly warns:
```
fn warn<T>(slice: &[T]) {
    let mut iter = slice.iter();
    loop {
        let _ = match iter.next() {
            Some(ele) => ele,
            None => break
        };
        //{}
    }
}
```

### PR DESCRIPTION
**Pull request by Wafflespeanut**
This should probably fix #359 (and I've updated some comments). But well, we should definitely rustup, because there's no way to test it right now :/

### CODE REVIEW

File: src/loops.rs [hunk 1]
```diff
@@ -310,9 +312,9 @@ fn is_iterable_array(ty: ty::Ty) -> bool {
 
 /// If a block begins with a statement (possibly a `let` binding) and has an expression, return it.
 fn extract_expr_from_first_stmt(block: &Block) -> Option<&Expr> {
-    match block.expr {
-        Some(_) => None,
-        None if !block.stmts.is_empty() => match block.stmts[0].node {
+    match block.stmts.is_empty() {
```
Comment (Manishearth) [inline src/loops.rs]: This should be an if statement

**Comment by Manishearth** [comment]
Rustup happened. There are type errors here.

Furthermore, I don't think this PR is changing anything in the detection, just the reporting (the bug is in the detection). I'll have to have a closer look later.

File: tests/compile-fail/while_loop.rs [hunk 2]
```diff
@@ -32,6 +32,8 @@ fn main() {
         };
         let _x = x;
         let _str = "foo";
+        { let _a = "bar"; };
+        { let _b = "foobar"; }
```
Comment (Manishearth) [inline tests/compile-fail/while_loop.rs]: Don't change the testcase, add another :smile:

File: src/loops.rs (hunk 1 above)
Comment (Wafflespeanut) [inline src/loops.rs]: C'mon, `match` looks beautiful!

**Comment by Wafflespeanut** [comment]
@Manishearth Initially, the `extract_expr_from_first_stmt` returned an expression (from the first statement) only if `block.expr` has `None` (which seems like the bug to me). So, once we have an inner block `{ }` like so, `block.expr` has `Some<P<Expr>>` in the AST, which has made clippy to ignore the `let` binding, and thereby moving to `extract_first_expr` which extracts the `{ }` out, and since it doesn't contain any `match`, it's also ignored.

Now that we return the first "thing" that exists in `block.stmts`, we always get an expression. So, it should work.

**Comment by Manishearth** [comment]
Ah, I see. Carry on then.

**Comment by Wafflespeanut** [comment]
@Manishearth We get one last lint error in `tests/compile-fail/while_loop.rs`. I'm not sure whether we should modify that regression test to allow the lint, or modify the loop?

**Comment by llogiq** [comment]
Fwiw, I think we should allow the lint – unless I misunderstand the code, the suggestion is correct.

**Comment by Manishearth** [comment]
Yes, you should mark it as allow since it's a regression test for something else.

**Comment by Manishearth** [comment]
Actually, just add an `//~ ERROR` there, since it should be emitting the error

**Comment by Wafflespeanut** [comment]
@Manishearth r?

File: src/loops.rs (hunk 1 above)
Comment (llogiq) [inline src/loops.rs]: `if block.stmts.is_empty() { None } else { match block.stmts[0].node { .. }` looks clearer. Or if I'm iterator-happy, even `block.stmts.iter().next().map(|stmt| match stmt.node { ... })`.

Comment (Wafflespeanut) [inline src/loops.rs]: @llogiq oh nice, will do :)

**Comment by Manishearth** [comment]
Needs rebase

**Comment by Wafflespeanut** [comment]
I'm sorry, internet problem, I was unable to do anything :/

**Comment by llogiq** [comment]
np, and your welcome @Wafflespeanut

**Pull request merged**